### PR TITLE
AWS Policies

### PR DESCRIPTION
**Pull request by tjholm**
Apply policies provided by config as code as part of AWS deployments.

### CODE REVIEW

File: pkg/stack/types.go [hunk 1]
```diff
@@ -136,6 +137,12 @@ type Stack struct {
 	Schedules   map[string]Schedule    `yaml:"schedules,omitempty"`
 	ApiDocs     map[string]*openapi3.T `yaml:"-"`
 	Apis        map[string]string      `yaml:"apis,omitempty"`
+	// TODO: Not currently supported by nitric.yaml configuration (but is technically definable using the proto model)
+	// We may want to decouple the definition from contracts at a later stage
+	// but re-using the contract here provides us a serializable entity with no
+	// repetition/redefinition
+	// NOTE: if we want to use the proto definition here we would need support for yaml parsing to use customisable tags
```
Comment (tjholm) [inline pkg/stack/types.go:144]: Not sure if we would want to docouple this definition from the contract policies. I assumed these would end up being much the same anyway in the nitric.yaml.

File: pkg/stack/types.go [hunk 2]
```diff
@@ -136,6 +137,12 @@ type Stack struct {
...
+	Policies []*v1.PolicyResource `yaml:"-"`
```
Comment (tjholm) [inline pkg/stack/types.go:146]: Not sure if we would want to docouple this definition from the contract policies. I assumed these would end up being much the same anyway in the nitric.yaml.

File: pkg/codeconfig/codeconfig.go [hunk 3]
```diff
@@ -429,14 +435,14 @@ func (c *codeConfig) ToStack() (*stack.Stack, error) {
...
 		if !ok {
-			f = stack.FunctionFromHandler(handler, s.Dir)
+			fun = stack.FunctionFromHandler(handler, s.Dir)
 		}
-		f.ComputeUnit.Triggers = stack.Triggers{
+		fun.ComputeUnit.Triggers = stack.Triggers{
 			Topics: topicTriggers,
 		}
-		s.Functions[rt.ContainerName()] = f
+		s.Functions[rt.ContainerName()] = fun
```
Comment (asalkeld) [inline pkg/codeconfig/codeconfig.go]: could use "name" here ^

File: pkg/provider/pulumi/aws/aws.go [hunk 4]
```diff
@@ -209,6 +236,24 @@ func (a *awsProvider) Deploy(ctx *pulumi.Context) error {
...
 	}
 
+	for _, p := range a.s.Policies {
+		policyName, err := policyResourceName(p)
+
+		if err != nil {
+			return err
+		}
+
+		newPolicy(ctx, policyName, &PolicyArgs{
```
Comment (asalkeld) [inline pkg/provider/pulumi/aws/aws.go]: _, err = newPolicy(...

File: pkg/stack/types.go (hunk 2 above)
Comment (asalkeld) [reply pkg/stack/types.go:146]: why not just have the FromOpts() populate this with the defaults (max allowed for each resource)

Comment (asalkeld) [reply pkg/stack/types.go:146]: haha, you have the skeleton below - just needs filling in else loading from file will break

Comment (tjholm) [reply pkg/stack/types.go:146]: Yeah put that in so we could just populate default actions for backwards compatibility (Just need to fill it in as you say :)).

This was more a question of whether we wanted to use our auto generated policy definitions from our proto contracts as part of the yaml definition.

eventually we could end up with something like:

```yaml
policies:
  principals:
     - name: test
       type: function
   resources:
     - name: test-topic
       type: topic
   actions:
      # Will want to use a custom deserializer here so we can use strings
      - 400
```

For users who would prefer to use a configuration file. This should almost work as is if we update the tags, howver I'm not sure if we can customise the tags for the yaml lib we're using to also use json tags (which are included on our proto generated sources).

File: pkg/codeconfig/codeconfig.go (hunk 3 above)
Comment (tjholm) [reply pkg/codeconfig/codeconfig.go]: fixed.

File: pkg/provider/pulumi/aws/aws.go (hunk 4 above)
Comment (tjholm) [reply pkg/provider/pulumi/aws/aws.go]: fixed

File: pkg/stack/types.go (hunk 2 above)
Comment (tjholm) [reply pkg/stack/types.go:146]: @asalkeld This should be sorted now

**Review by asalkeld — APPROVED** [review]

**Pull request merged**